### ISSUE TITLE
Problems using nested case classes

### ISSUE THREAD

**Issue by Knorreman** [vincenzobaz/spark-scala3#44]
Hello

We need to use nested case classes to model our data. 

    import org.apache.spark.sql.catalyst.encoders.encoderFor
    import scala3encoders.given

    case class Origin(countries: Seq[String])
    case class Vegetables(name: String, origin: Origin)
    val encoder = encoderFor[Vegetables]

This causes a compile error

    No given instance of type scala3encoders.derivation.Deserializer[Origin] was found.
    I found:

    scala3encoders.derivation.Deserializer.derivedProduct[Origin](
      Origin.$asInstanceOf[
        
          scala.deriving.Mirror.Product{
            type MirroredMonoType = Origin; type MirroredType = Origin;
              type MirroredLabel = ("Origin" : String);
              type MirroredElemTypes = Seq[String] *: EmptyTuple.type;
              type MirroredElemLabels = ("countries" : String) *: EmptyTuple.type
          }
        
      ],
    scala.reflect.ClassTag.apply[Origin](classOf[Origin]))

    But given instance derivedProduct in object Deserializer does not match type scala3encoders.derivation.Deserializer[Origin].
        val encoder = encoderFor[Vegetables]

Am I using the encoder wrong or is it something else I should do?

**Comment by vincenzobaz** [comment]
Hello @Knorreman I suspect the issue might be with `encoderFor`.
If you try with 
```
summon[Encoder[Vegetables]]
```
it should work

**Comment by Knorreman** [comment]
Hi @vincenzobaz 
Unfortunately I get exactly the same error with
`summon[Encoder[Vegetables]]`

**Comment by vincenzobaz** [comment]
Are you using version `0.2.3`?

This program
```
import org.apache.spark.sql.Encoder

object Vegetables extends App:
  import scala3encoders.given

  case class Origin(countries: Seq[String])
  case class Vegetables(name: String, origin: Origin)
  val encoder = summon[Encoder[Vegetables]]
  println(encoder.schema)
```
prints 
```
[info] StructType(StructField(name,StringType,true),StructField(origin,StructType(StructField(countries,ArrayType(StringType,true),true)),true))
```
for me

**Comment by Knorreman** [comment]
Yes I am using 0.2.3
But I was using spark 3.4.1 and when I tried 3.3.2 it worked!

**Comment by vincenzobaz** [comment]
Cool! Try also spark 3.5.0

**Comment by vincenzobaz** [comment]
Cool! It is the version that the library is based. We have just merged compatibility for 3.5.0 but it is not released yet

**Comment by michael72** [comment]
so - the main reason for this error in 0.2.3 seems to happen once you `summon[Encoder[Seq[String]]]` then we get

```
val bla = summon[Encoder[Origin]]
[error]     |                                     ^
[error]     |No given instance of type scala3encoders.derivation.Deserializer[Seq[String]] was found.
[error]     |I found:
[error]     |
[error]     |    scala3encoders.derivation.Deserializer.deriveSeq[Seq, String](
[error]     |      scala3encoders.derivation.Deserializer.given_Deserializer_String,
[error]     |      scala.reflect.ClassTag.apply[String](classOf[String]))(<:<.refl[Seq[String]])
[error]     |
[error]     |But given instance deriveSeq in object Deserializer does not match type scala3encoders.derivation.Deserializer[Seq[String]].
```

But this was also the part that was reworked in spark 3.5.0 and I seemed to have taken that over with the new Helper class... With a `publishLocal` of a 0.2.4 version it worked for me.

**Comment by Knorreman** [comment]
Nice! Any ideas when 0.2.4 will be published? We can downgrade to 3.3.2 in the meantime! Thanks guys!

**Comment by vincenzobaz** [comment]
I am running the release now.
